compiler/src/iree/compiler/Codegen/Dialect/MicroKernelOps.cpp

```diff
@@ -0,0 +1,324 @@
+          static_cast<int64_t>(outsPosAndSize.first + outsPosAndSize.second)};
+}
+
+/// Map type of operand of a `generic_micro_kernel` operation to the type(s) of
+/// the function call arguments(s) it lowers to.
+static LogicalResult getCallArgumentTypeForOperand(
+    MLIRContext *context, Type microKernelOpOperandType,
+    SmallVector<Type> &callOperandTypes) {
+  return TypeSwitch<Type, LogicalResult>(microKernelOpOperandType)
+      .Case<FloatType, IndexType, IntegerType>([&](auto scalarType) {
+        callOperandTypes.push_back(scalarType);
+        return success();
+      })
+      .Case<MemRefType>([&](MemRefType memrefType) {
+        // 0D memref lowers to a 0D memref operand type. Other nD memrefs
+        // lower to 0D memref, offset, and n-1 strides.
+
+        // Base ptr.
+        callOperandTypes.push_back(MemRefType::get(
+            ArrayRef<int64_t>{}, memrefType.getElementType(),
+            MemRefLayoutAttrInterface{}, memrefType.getMemorySpace()));
+        if (memrefType.getRank() == 0) return success();
+
+        auto indexType = IndexType::get(context);
+        // Offset
+        callOperandTypes.push_back(indexType);
+
+        // Strides.
+        callOperandTypes.resize(
+            callOperandTypes.size() + memrefType.getRank() - 1, indexType);
+        return success();
+      })
+      .Default([&](Type t) { return failure(); });
+}
+
+/// Map `operand` of a `generic_micro_kernel` operation to the operand(s) of
+/// the function call it lowers to.
+static LogicalResult lowerToCallOperands(Location loc, RewriterBase &rewriter,
+                                         Value operand,
+                                         SmallVector<Value> &callOperands) {
+  return TypeSwitch<Type, LogicalResult>(operand.getType())
+      .Case<FloatType, IndexType, IntegerType>([&](auto scalarType) {
+        callOperands.push_back(operand);
+        return success();
+      })
+      .Case<MemRefType>([&](MemRefType memrefType) {
+        auto extractStridedMetadataOp =
+            rewriter.create<memref::ExtractStridedMetadataOp>(loc, operand);
+        // Base ptr.
+        callOperands.push_back(extractStridedMetadataOp.getBaseBuffer());
+        if (memrefType.getRank() == 0) {
+          return success();
+        }
+        // Offset.
+        callOperands.push_back(extractStridedMetadataOp.getOffset());
+        // Strides.
+        for (auto stride : extractStridedMetadataOp.getStrides().drop_back()) {
+          callOperands.push_back(stride);
+        }
+        return success();
+      })
+      .Default([](Type) { return failure(); });
+}
+
+FailureOr<func::CallOp> GenericMicroKernelOp::lowerToFunctionCall(
+    RewriterBase &rewriter) {
+  // TODO handle op with return values if they are scalar.
+  if (getNumResults() != 0) {
+    return rewriter.notifyMatchFailure(
+        getOperation(), "cannot lower to function call operation with results");
+  }
```

Comment:
nit: can you add your username to the TODO? That would help others know that you're the person with the best details about this.

just a note: my understanding about the username in TODO is that a TODO is not a commitment that the person referenced will fix the problem. It is more how it can be searched to find out how to get more details upon request.

https://google.github.io/styleguide/cppguide.html#TODO_Comments

Reply:
I like our current practice of `//TODO(#12345)` --- it forces you to file an issue and dump your current brain context, which always feel like not much at the time you're doing it, but would take others in the future some effort to reconstruct if you don't save it now.

Reply:
Or I can just handle it :)